### PR TITLE
(devbreak) disallow use of empty table metadata

### PR DESCRIPTION
**Pull request by clockfort**
**Goals (and why)**:
We have some weird error output in $DOWNSTREAM_PROJECT after I merged #1822 that would indicate that despite me playing whack-a-mole and thinking I found all the places that tests downstream that used empty table metadata, it looks like some ETE tests do something more insidious that I didn't catch. (the new #1822 code doesn't actually fail, but the cache warming fails because KVS.getMetadataForTables() returns some empty byte arrays which it can't deserialize, which Should Never Happen™)

Adding this guarantee/throw should hopefully let me catch the offending tests at metadata table insert time, and we should make this guarantee anyways, for our own sanity's sake, as tables with empty metadata leave too many questions for us with how to deal with them. (what conflict handling should they have? etc)

** Concerns **:
I'm expecting (hoping!) this breaks some of $BIG_DOWNSTREAM_PROJECT's tests; it may also break some of our other downstream projects' tests if they were the same variety of lazy when writing tests.
I'm okay with that, and they should switch to the real table builder or use AtlasDbConstants.GENERIC_TABLE_METADATA if they want to continue being lazy.

**Priority (whenever / two weeks / yesterday)**:
two weeks; it's a test only problem and it's not actually breaking any tests, just cluttering the error output and confusing people looking for other issues

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/1925)

### CODE REVIEW

File: atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/ValidatingQueryRewritingKeyValueService.java [hunk 1]
```diff
@@ -104,6 +105,15 @@ protected void sanityCheckTableName(TableReference tableRef) {
                 "invalid tableName: " + tableName);
     }
 
+    protected void sanityCheckTableMetadata(TableReference tableRef, byte[] tableMetadata) {
+        if (tableMetadata == null || Arrays.equals(tableMetadata, AtlasDbConstants.EMPTY_TABLE_METADATA)) {
+            throw new IllegalArgumentException(
+                    String.format(
+                            "Passing in empty table metadata for table '%s' does not make sense, and is disallowed.",
```
Comment (gsheasby) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/ValidatingQueryRewritingKeyValueService.java]: This message could be less snarky :)
Could also recommend using `AtlasDbConstants.GENERIC_TABLE_METADATA`?

File: docs/source/release_notes/release-notes.rst [hunk 2]
```diff
@@ -81,6 +81,12 @@ develop
          - Fixed a bug with import ordering and license generation in IntelliJ not respecting Baseline conventions.
            (`Pull Request <https://github.com/palantir/atlasdb/pull/1893>`__)
 
+    *    - |devbreak|
+         - Some downstream projects were using empty table metadata for dev-laziness reasons in their tests.
+           This is no longer permitted, as it leads to many (unsolved) questions about how to deal with such a table.
+           If this breaks your tests, you can fix it with making real schema for tests or switch to AtlasDbConstants.GENERIC_TABLE_METADATA
```
Comment (gsheasby) [inline docs/source/release_notes/release-notes.rst]: grammatical nit: Please switch `switch` for `switching`.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/ValidatingQueryRewritingKeyValueService.java [hunk 3]
```diff
@@ -92,6 +92,7 @@ public void createTables(Map<TableReference, byte[]> tableRefToTableMetadata) {
         for (TableReference tableRef : tableRefToTableMetadata.keySet()) {
             sanityCheckTableName(tableRef);
         }
+        tableRefToTableMetadata.entrySet().forEach(entry -> sanityCheckTableMetadata(entry.getKey(), entry.getValue()));
```
Comment (gsheasby) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/ValidatingQueryRewritingKeyValueService.java]: Would you like to do a drive-by java-8-ifying of the above for loop, too?

**Comment by clockfort** [comment]
made changes as per PR and rebased onto latest develop to fix release notes merge conflict
